etna/datasets/tsdataset.py

```diff
@@ -708,6 +709,72 @@ def to_dataset(df: pd.DataFrame) -> pd.DataFrame:
         df_copy = df_copy.sort_index(axis=1, level=(0, 1))
         return df_copy
 
+    @staticmethod
+    def _hierarchical_structure_from_level_columns(
+        df: pd.DataFrame, level_columns: List[str], sep: str
+    ) -> HierarchicalStructure:
+        """Create hierarchical structure from dataframe columns."""
+        df_level_columns = df[level_columns]
+
+        prev_level_name = level_columns[0]
+        for cur_level_name in level_columns[1:]:
+            df_level_columns[cur_level_name] = (
+                df_level_columns[prev_level_name] + sep + df_level_columns[cur_level_name]
+            )
+            prev_level_name = cur_level_name
+
+        level_structure = {"total": df[level_columns[0]].unique()}
+        cur_level_name = level_columns[0]
+        for next_level_name in level_columns[1:]:
+            cur_level_to_next_level_edges = df_level_columns[[cur_level_name, next_level_name]].drop_duplicates()
+            cur_level_to_next_level_adjacency_list = cur_level_to_next_level_edges.groupby(cur_level_name).agg(list)
+            level_structure.update(cur_level_to_next_level_adjacency_list.to_records())
+
+        hierarchical_structure = HierarchicalStructure(
+            level_structure=level_structure, level_names=["total"] + level_columns
+        )
+        return hierarchical_structure
+
+    @staticmethod
+    def to_hierarchical_dataset(
```

Comment:
We can probably also look at [overload](https://docs.python.org/3/library/typing.html#typing.overload) for highlighting dependency of return type on flag `return_hierarchy`.

Reply:
Don't get how do you suggest to do it?

Reply:
We can try to make two signatures: with `return_hierarchy=True` and `return_hierarchy=False` and different output types.